File: android/core-transport/src/main/kotlin/dev/omnesis/android/transport/dto/SourceNotice.kt
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Adrien Conrath

package dev.omnesis.android.transport.dto

import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonDecoder
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import java.time.Instant
import java.time.OffsetDateTime
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.Locale

/**
 * One thing a person should be told about a source on one device, worded by the
 * gateway. Mirrors `SourceNotice` in packages/types/src/source-notice.ts.
 *
 * Clients render [title], [detail] and [steps] verbatim and never derive notices
 * from `errorMessage`, `issues` or `coverage` themselves. The one exception is
 * [SourceSyncStatus.displayNotices], which stands in for a gateway that predates
 * the field (see [fallbackNotices]). [kind] and [severity] stay strings so a value
 * this build does not know still decodes; [level] maps an unknown severity to
 * [NoticeLevel.WARNING]. Lists decode through [LenientNoticeListSerializer], so one
 * malformed notice is skipped rather than failing the status it rides on.
 */
@Serializable
data class SourceNotice(
    val kind: String = "",
    val severity: String = "warning",
    val title: String = "",
    val detail: String? = null,
    val steps: List<String>? = null,
    /** ISO 8601 time the condition was first observed, when the gateway knows it. */
    val since: String? = null,
) {
    val level: NoticeLevel
        get() = when (severity) {
            "info" -> NoticeLevel.INFO
            "error" -> NoticeLevel.ERROR
            else -> NoticeLevel.WARNING
        }
}

/** A notice's severity, ordered from least to most severe. */
enum class NoticeLevel { INFO, WARNING, ERROR }

/**
 * Decodes a notice list one element at a time. An element that is not an object or
 * has no non-blank title is dropped, non-string fields read as absent, and null steps
 * are skipped — a notice is advisory, so a bad one must never cost the whole
 * `/admin/sync/status` page.
 */
object LenientNoticeListSerializer : KSerializer<List<SourceNotice>> {
    private val delegate = ListSerializer(SourceNotice.serializer())
    override val descriptor: SerialDescriptor = delegate.descriptor

    override fun serialize(encoder: Encoder, value: List<SourceNotice>) = delegate.serialize(encoder, value)

    override fun deserialize(decoder: Decoder): List<SourceNotice> {
        val json = decoder as? JsonDecoder ?: return delegate.deserialize(decoder)
        val array = json.decodeJsonElement() as? JsonArray ?: return emptyList()
        return array.mapNotNull { element -> (element as? JsonObject)?.let(::noticeFrom) }
    }

    private fun noticeFrom(o: JsonObject): SourceNotice? {
        fun str(key: String): String? = (o[key] as? JsonPrimitive)?.takeIf { it.isString }?.content
        val title = str("title")?.takeIf(String::isNotBlank) ?: return null
        val steps = (o["steps"] as? JsonArray)
            ?.mapNotNull { step -> (step as? JsonPrimitive)?.takeIf { it.isString }?.content }
        return SourceNotice(
            kind = str("kind").orEmpty(),
            severity = str("severity") ?: "warning",
            title = title,
            detail = str("detail"),
            steps = steps,
            since = str("since"),
        )
    }
}

/** Titles of the notices synthesized for a gateway that predates `notices`. */
object FallbackNoticeTitles {
    const val NEEDS_AUTH = "Needs sign-in"
    const val RATE_LIMITED = "Paused by the provider's rate limit"
    const val ERROR = "The last sync failed"
    const val STALE = "No new data is arriving"
    const val EXPIRING_PREFIX = "Connection expires on "
}

/**
 * The notices an older gateway would have sent for [status], derived from its state:
 *
 * - `needs-auth` → error "Needs sign-in", detail = `errorMessage` without "needs reauth: ";
 * - `rate-limited` → info "Paused by the provider's rate limit", detail without "rate-limited: ";
 * - `error` → error "The last sync failed", detail = `errorMessage`;
 * - `stale` with a `staleHint` → warning "No new data is arriving", detail = the hint;
 * - `auth-expiring` with a `consentExpiresAt` → warning "Connection expires on <date>".
 *
 * Used only when `notices` is absent. The same table lives in the iOS client and
 * mirrors the gateway's wording, so an older gateway reads the same as a current one.
 */
internal fun fallbackNotices(status: SourceSyncStatus): List<SourceNotice> {
    val message = status.errorMessage?.takeIf(String::isNotBlank)
    val notice = when (status.state) {
        "needs-auth" -> SourceNotice(
            kind = "needs-auth", severity = "error", title = FallbackNoticeTitles.NEEDS_AUTH,
            detail = message?.removePrefix("needs reauth: ")?.takeIf(String::isNotBlank),
        )
        "rate-limited" -> SourceNotice(
            kind = "rate-limited", severity = "info", title = FallbackNoticeTitles.RATE_LIMITED,
            detail = message?.removePrefix("rate-limited: ")?.takeIf(String::isNotBlank),
        )
        "error" -> SourceNotice(kind = "error", severity = "error", title = FallbackNoticeTitles.ERROR, detail = message)
        "stale" -> status.staleHint?.takeIf(String::isNotBlank)?.let { hint ->
            SourceNotice(kind = "stale", severity = "warning", title = FallbackNoticeTitles.STALE, detail = hint)
        }
        "auth-expiring" -> status.consentExpiresAt?.takeIf(String::isNotBlank)?.let { deadline ->
            SourceNotice(
                kind = "auth-expiring",
                severity = "warning",
                title = FallbackNoticeTitles.EXPIRING_PREFIX + expiryDate(deadline),
            )
        }
        else -> null
    }
    return listOfNotNull(notice)
}

private val EXPIRY_DATE = DateTimeFormatter.ofPattern("MMM d, yyyy", Locale.US)

/** "Jul 15, 2026" in the device's zone, or the raw value when it does not parse. */
private fun expiryDate(iso: String): String {
    val instant = runCatching { Instant.parse(iso) }.getOrNull()
        ?: runCatching { OffsetDateTime.parse(iso).toInstant() }.getOrNull()
        ?: return iso
    return EXPIRY_DATE.withZone(ZoneId.systemDefault()).format(instant)
}

File: ios/Sources/Omnesis/Transport/SourceNotice.swift
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Adrien Conrath

import Foundation

/// One thing a person should be told about a source on one device. Mirrors
/// `SourceNotice` in packages/types/src/source-notice.ts. The gateway writes
/// every word; clients present them and never interpret the underlying
/// status fields themselves.
public struct SourceNotice: Decodable, Equatable, Hashable, Sendable {
    /// How loudly to present a notice. Ordered from least to most severe, so
    /// `max()` picks the one a single indicator should show.
    public enum Severity: String, Comparable, Sendable, CaseIterable {
        case info
        case warning
        case error

        private var rank: Int {
            switch self {
            case .info: 0
            case .warning: 1
            case .error: 2
            }
        }

        public static func < (lhs: Severity, rhs: Severity) -> Bool {
            lhs.rank < rhs.rank
        }

        /// The noun for a count of notices at this severity, as the portal
        /// words it: problems, warnings and notes.
        public func noun(count: Int) -> String {
            switch self {
            case .info: count == 1 ? "note" : "notes"
            case .warning: count == 1 ? "warning" : "warnings"
            case .error: count == 1 ? "problem" : "problems"
            }
        }
    }

    /// Stable category (`error`, `needs-auth`, `coverage-partial`, …). Kept as
    /// a string so a kind this build has never seen still decodes.
    public let kind: String
    /// A severity this build does not know decodes as `.warning`: shown, never
    /// dropped, never alarming.
    public let severity: Severity
    public let title: String
    public let detail: String?
    public let steps: [String]?
    /// ISO 8601 time the condition was first observed.
    public let since: String?

    public init(
        kind: String,
        severity: Severity,
        title: String,
        detail: String? = nil,
        steps: [String]? = nil,
        since: String? = nil
    ) {
        self.kind = kind
        self.severity = severity
        self.title = title
        self.detail = detail
        self.steps = steps
        self.since = since
    }

    private enum CodingKeys: String, CodingKey {
        case kind
        case severity
        case title
        case detail
        case steps
        case since
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        kind = try container.decode(String.self, forKey: .kind)
        let rawSeverity = try container.decodeIfPresent(String.self, forKey: .severity)
        severity = rawSeverity.flatMap(Severity.init(rawValue:)) ?? .warning
        title = try container.decode(String.self, forKey: .title)
        detail = try container.decodeIfPresent(String.self, forKey: .detail)
        steps = try container.decodeIfPresent([String].self, forKey: .steps)
        since = try container.decodeIfPresent(String.self, forKey: .since)
    }
}

/// The notices one device has for a source. `deviceId` is `nil` when the
/// gateway named no device for them.
public struct DeviceNotices: Equatable, Sendable {
    public let deviceId: String?
    public let notices: [SourceNotice]

    public init(deviceId: String?, notices: [SourceNotice]) {
        self.deviceId = deviceId
        self.notices = notices
    }
}

extension SourceSyncStatus {
    /// The notices to show for this status. A gateway that serves `notices`
    /// is taken at its word, including an empty list. For one that does not —
    /// an older gateway, or a status the socket has just moved into a new
    /// state before the gateway was read again — the notice is derived from
    /// the state itself, in the gateway's own words, so a failure or warning
    /// is never left unmentioned.
    public var displayNotices: [SourceNotice] {
        if let notices { return notices }
        return fallbackNotices
    }

    private var fallbackNotices: [SourceNotice] {
        let message = errorMessage.flatMap { $0.isEmpty ? nil : $0 }
        switch state {
        case "needs-auth":
            return [SourceNotice(
                kind: "needs-auth",
                severity: .error,
                title: "Needs sign-in",
                detail: message.flatMap { strippingPrefix("needs reauth: ", from: $0) }
            )]
        case "rate-limited":
            return [SourceNotice(
                kind: "rate-limited",
                severity: .info,
                title: "Paused by the provider's rate limit",
                detail: message.flatMap { strippingPrefix("rate-limited: ", from: $0) }
            )]
        case "stale" where staleHint?.isEmpty == false:
            return [SourceNotice(kind: "stale", severity: .warning, title: "No new data is arriving", detail: staleHint)]
        case "auth-expiring" where consentExpiresAt != nil:
            let title = parseISODate(consentExpiresAt).map {
                "Connection expires on \($0.formatted(.dateTime.day().month(.wide).year()))"
            } ?? "Connection expires soon"
            return [SourceNotice(kind: "auth-expiring", severity: .warning, title: title)]
        default:
            guard state == "error" || message != nil else { return [] }
            return [SourceNotice(kind: "error", severity: .error, title: "The last sync failed", detail: message)]
        }
    }

    private func strippingPrefix(_ prefix: String, from message: String) -> String? {
        let stripped = message.hasPrefix(prefix) ? String(message.dropFirst(prefix.count)) : message
        return stripped.isEmpty ? nil : stripped
    }

    /// The notices a status keeps when a socket event updates it. Notices
    /// never ride an event, so the last ones read are kept — unless the event
    /// changed the state or the failure, when they may describe the old state.
    /// Then there are none (`nil`), and `displayNotices` derives one from the
    /// new state until the gateway is read again.
    static func carriedNotices(
        from existing: SourceSyncStatus?,
        state: String,
        errorMessage: String?
    )
        -> [SourceNotice]? {
        guard let existing, existing.state == state, existing.errorMessage == errorMessage else { return nil }
        return existing.notices
    }

    /// A copy with the notices of a fresher read of the same source. Every
    /// other field is kept, so a status the socket has moved on since the read
    /// was taken keeps its newer state. Members are matched by device; one the
    /// read does not list has nothing to say. When the read has members this
    /// copy lacks, their rows are taken whole, since there is no newer
    /// per-device state to keep.
    func adoptingNotices(from fresh: SourceSyncStatus) -> SourceSyncStatus {
        let updatedMembers: [SourceSyncStatus]? = if let members, !members.isEmpty {
            members.map { member in
                guard let freshMember = fresh.members?.first(where: { $0.deviceId == member.deviceId }) else {
                    return member.withNotices([])
                }
                return member.withNotices(freshMember.notices)
            }
        } else {
            fresh.members
        }
        return replacingMembers(updatedMembers).withNotices(fresh.notices)
    }

    private func withNotices(_ notices: [SourceNotice]?) -> SourceSyncStatus {
        SourceSyncStatus(
            sourceId: sourceId,
            deviceId: deviceId,
            members: members,
            state: state,
            unitName: unitName,
            progress: progress,
            startedAt: startedAt,
            lastSyncAt: lastSyncAt,
            errorMessage: errorMessage,
            erroredAt: erroredAt,
            lastUpdated: lastUpdated,
            consentExpiresAt: consentExpiresAt,
            staleHint: staleHint,
            notices: notices
        )
    }

    /// Adopt the notices of a fresh read into every status already held.
    /// Sources the read adds or drops are left to the full refresh.
    static func adoptingNotices(
        from fresh: [SourceSyncStatus],
        into current: [String: SourceSyncStatus]
    )
        -> [String: SourceSyncStatus] {
        var updated = current
        for status in fresh {
            guard let held = current[status.sourceId] else { continue }
            updated[status.sourceId] = held.adoptingNotices(from: status)
        }
        return updated
    }

    /// The notices for each device that contributes to this source. A
    /// multi-device status reads each member's own list; any other status
    /// belongs wholly to one device — its own `deviceId`, or
    /// `fallbackDeviceId` (the source's registered host) when it names none.
    public func noticesByDevice(fallbackDeviceId: String?) -> [DeviceNotices] {
        if let members, !members.isEmpty {
            return members.map { DeviceNotices(deviceId: $0.deviceId, notices: $0.displayNotices) }
        }
        return [DeviceNotices(deviceId: deviceId ?? fallbackDeviceId, notices: displayNotices)]
    }

    /// Every notice across the source's devices, for a surface with room for
    /// only one indicator.
    public var allDisplayNotices: [SourceNotice] {
        if let members, !members.isEmpty {
            return members.flatMap(\.displayNotices)
        }
        return displayNotices
    }
}

/// A notice list ordered most severe first, keeping the gateway's order
/// within a severity.
func sortedBySeverity(_ notices: [SourceNotice]) -> [SourceNotice] {
    notices.enumerated()
        .sorted { lhs, rhs in
            lhs.element.severity != rhs.element.severity
                ? lhs.element.severity > rhs.element.severity
                : lhs.offset < rhs.offset
        }
        .map(\.element)
}

/// The label for one severity's icon: "2 warnings for studio-desk".
func sourceNoticeGroupLabel(severity: SourceNotice.Severity, count: Int, deviceName: String?) -> String {
    let base = "\(count) \(severity.noun(count: count))"
    guard let deviceName, !deviceName.isEmpty else { return base }
    return "\(base) for \(deviceName)"
}

/// The label for a single icon standing for many notices: "3 notices,
/// including 1 problem, for studio-desk and travel-laptop". When every notice
/// shares a severity, the count names it: "2 warnings".
func sourceNoticeSummaryLabel(_ notices: [SourceNotice], deviceNames: [String]) -> String {
    guard let worst = notices.map(\.severity).max() else { return "" }
    let worstCount = notices.filter { $0.severity == worst }.count
    var label = if worstCount == notices.count {
        "\(notices.count) \(worst.noun(count: notices.count))"
    } else {
        "\(notices.count) notices, including \(worstCount) \(worst.noun(count: worstCount))"
    }
    let names = deviceNames.filter { !$0.isEmpty }
    if !names.isEmpty {
        label += "\(worstCount == notices.count ? "" : ",") for \(ListFormatter.localizedString(byJoining: names))"
    }
    return label
}

/// "Since 2 Jan, 03:04" — the portal's phrasing, in the reader's locale.
func noticeSinceText(_ iso: String?) -> String? {
    guard let date = parseISODate(iso) else { return nil }
    return "Since \(date.formatted(.dateTime.day().month(.abbreviated).hour().minute()))"
}

/// Parse an ISO 8601 timestamp, with or without fractional seconds
/// (`ISO8601DateFormatter` accepts only the form its options name, so both
/// are tried).
func parseISODate(_ iso: String?) -> Date? {
    guard let iso, !iso.isEmpty else { return nil }
    let withFractional = ISO8601DateFormatter()
    withFractional.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    if let date = withFractional.date(from: iso) { return date }
    let plain = ISO8601DateFormatter()
    plain.formatOptions = [.withInternetDateTime]
    return plain.date(from: iso)
}
